Enable external signer support by default, reduce #ifdef

Sjors opened the pull request:
This follows the introduction of GUI support in https://github.com/bitcoin-core/gui/pull/4

I don't think we should expect GUI users to self compile. This also enables external signer support by default for RPC users.

In addition this PR reduces the number of `#ifdef ENABLE_EXTERNAL_SIGNER`, which also fixes #21919. When compiled with `--disable-external-signer` such wallets can't be created in RPC or GUI, but they can be loaded. Attempting any action that calls HWI will trigger an error.

Side-note: this PR may or may not (currently) break CI for the GUI repository, as explained here: https://github.com/bitcoin-core/gui/pull/4#issuecomment-769859001

Sjors committed:
build: enable external signer by default

ryanofsky commented on src/interfaces/node.h:
```diff
@@ -110,6 +111,11 @@ class Node
     //! Disconnect node by id.
     virtual bool disconnectById(NodeId id) = 0;
 
+#ifdef ENABLE_EXTERNAL_SIGNER
```
Would be nice to drop `#ifdef ENABLE_EXTERNAL_SIGNER` in the interface here and just return an empty vector in the implementation or throw a not-implemented exception. Cross-process interfaces are trying not to be conditional to avoid build complexity and fragile incompatibilities between builds.

Also maybe in general in this PR, preprocessor is a little more widely used that it needs to be. Preprocessor is great when you need to write platform specific fragments or cut out a significant external or internal dependency like boost or the wallet. But if you just want to act conditionally, normal code is a good way to go. I'd try to avoid conditional compilation except for the parts of the code directly depending on boost.

Sjors replied:
I added a commit to reduce `#ifdef ENABLE_EXTERNAL_SIGNER` usage, including in the interface.

Rspigler commented:
Some thoughts:

Arguments in favor:
Many people use HWWs. This will lead to an increase in Core full node use.
GUI users don't self-compile (and shouldn't be expected to)
Still opt-in for `bitcoind`
Advanced users can still compile GUI without

Arguments against:
Requires boost::process
Many security concerns were raised in this PR (https://github.com/bitcoin/bitcoin/pull/15382) - I can't tell if they were addressed fully enough to change the default to enabled (@practicalswift, @jonasschnelli, @laanwj, @MarcoFalke)

Sjors committed:
refactor: reduce #ifdef ENABLE_EXTERNAL_SIGNER usage

In particular this make the node interface independent on whether external signer support is compiled.

Sjors committed:
refactor: clean up external_signer.h includes

Co-Authored-By: Hennadii Stepanov <32963518+hebasto@users.noreply.github.com>

Sjors commented on src/wallet/wallet.cpp:
```diff
@@ -8,6 +8,7 @@
 #include <chain.h>
 #include <consensus/consensus.h>
 #include <consensus/validation.h>
+#include <external_signer.h>
```
edaff3958e3c9d92e7aa15265ecad155b05714bf: I might make a separate commit for this fix

Sjors commented:
The Win64 machine is failing to build the fuzz executable: `undefined reference to `ExternalSigner::GetDescriptors(int)'`. It's configured with `--disable-external-signer`, but that shouldn't matter...

Sjors commented:
Rebased after  bitcoin-core/gui#4 was merged.

fanquake commented:
[Windows build failure](https://github.com/bitcoin/bitcoin/pull/21935/checks?check_run_id=2786475113):
```bash
/usr/bin/x86_64-w64-mingw32-ld: libbitcoin_wallet.a(libbitcoin_wallet_a-wallet.o):wallet.cpp:(.text+0x1011f): undefined reference to `ExternalSigner::GetDescriptors(int)'
/usr/bin/x86_64-w64-mingw32-ld: libbitcoin_wallet.a(libbitcoin_wallet_a-external_signer_scriptpubkeyman.o):external_signer_scriptpubkeyman.cpp:(.text+0x422): undefined reference to `ExternalSigner::DisplayAddress(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const'
/usr/bin/x86_64-w64-mingw32-ld: libbitcoin_wallet.a(libbitcoin_wallet_a-external_signer_scriptpubkeyman.o):external_signer_scriptpubkeyman.cpp:(.text+0xf68): undefined reference to `ExternalSigner::Enumerate(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<ExternalSigner, std::allocator<ExternalSigner> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)'
/usr/bin/x86_64-w64-mingw32-ld: libbitcoin_wallet.a(libbitcoin_wallet_a-external_signer_scriptpubkeyman.o):external_signer_scriptpubkeyman.cpp:(.text+0x1496): undefined reference to `ExternalSigner::SignTransaction(PartiallySignedTransaction&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&)'
collect2: error: ld returned 1 exit status
```

Sjors commented:
@fanquake indeed. I was hoping a rebase would magically make it go away... https://github.com/bitcoin/bitcoin/pull/21935#issuecomment-840680585

meshcollider commented on src/wallet/external_signer_scriptpubkeyman.h:
```diff
@@ -32,5 +31,3 @@ class ExternalSignerScriptPubKeyMan : public DescriptorScriptPubKeyMan
 };
 #endif
-
-#endif // BITCOIN_WALLET_EXTERNAL_SIGNER_SCRIPTPUBKEYMAN_H
```
I think you should keep the comment for style

Sjors replied:
I'm surprised the linter didn't pick that up, fixed.

hebasto commented:
In the "build: enable external signer by default" commit all `--enable-external-signer` could be removed from CI tasks.

hebasto commented:
The link error could be fixed with the following patch:
```diff
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -399,6 +399,7 @@ endif
 libbitcoin_wallet_a_CPPFLAGS = $(AM_CPPFLAGS) $(BITCOIN_INCLUDES) $(SQLITE_CFLAGS)
 libbitcoin_wallet_a_CXXFLAGS = $(AM_CXXFLAGS) $(PIE_FLAGS)
 libbitcoin_wallet_a_SOURCES = \
+  external_signer.cpp \
   wallet/coincontrol.cpp \
   wallet/context.cpp \
   wallet/crypter.cpp \
@@ -539,7 +540,6 @@ libbitcoin_common_a_SOURCES = \
   compressor.cpp \
   core_read.cpp \
   core_write.cpp \
-  external_signer.cpp \
   init/common.cpp \
   key.cpp \
   key_io.cpp \
```

Although, not sure if it is a proper fix.

hebasto commented:
Here is another possible fixup for the link error:
```diff
--- a/src/Makefile.test.include
+++ b/src/Makefile.test.include
@@ -160,7 +160,7 @@ BITCOIN_TESTS += \
   wallet/test/scriptpubkeyman_tests.cpp
 
 FUZZ_SUITE_LD_COMMON +=\
- $(LIBBITCOIN_WALLET) \
+ $(LIBBITCOIN_WALLET) $(LIBBITCOIN_COMMON) \
  $(SQLITE_LIBS) \
  $(BDB_LIBS)
 
```

The choice between two suggested fixups depends on the decision which library the `external_signer` belongs to -- `libbitcoin_wallet` or `libbitcoin_common`.

Sjors commented:
@hebasto I tried your second approach. The ExternalSigner class is independent of the wallet, e.g. `enumerate` can be used and later on we can support signing stuff without a wallet (similar to `signrawtransaction`). But it doesn't help (`LIBBITCOIN_COMMON` is already part of `FUZZ_SUITE_LD_COMMON`).

I'm able to reproduce on Linux as well, like @achow101. Using something like:

```
 ./configure --without-gui --disable-external-signer --disable-bench
``` 

@achow101 I reverted your commit

Also rebased, because who doesn't like that?

> In the "build: enable external signer by default" commit all `--enable-external-signer` could be removed from CI tasks.

Done

hebasto commented:
@Sjors 

> @hebasto I tried your second approach. The ExternalSigner class is independent of the wallet, e.g. `enumerate` can be used and later on we can support signing stuff without a wallet (similar to `signrawtransaction`)
> 
> But that change didn't help (`LIBBITCOIN_COMMON` is already part of `FUZZ_SUITE_LD_COMMON`).

It is not about being a part, but about ordering.

> I'm able to reproduce on Linux as well, like @achow101. Using something like:
> 
> ```
>  ./configure --without-gui --disable-external-signer --disable-bench
> ```

Works fine for me with https://github.com/bitcoin/bitcoin/pull/21935#issuecomment-859701171

Sjors commented:
Odd, it indeed does work. Maybe I forgot to clean something the last time.

hebasto reviewed (approved):
ACK 718fdce93c83f971d5fc70fe53c74d8a2165381a, tested on Linux Mint 20.1 (Qt 5.12.8) with `--enable-external-signer` and `--disable-external-signer` configure options. #21919 is fixed.

In commit _"wallet: add missing external_signer.h include"_ (fe337a84652f3fd3a261d01587b0a2f032944ca1) including of the `<external_signer.h>` header could be improved:
- add to `node/interfaces.cpp`, `qt/walletcontroller.cpp`
- drop in `wallet/scriptpubkeyman.cpp`, `wallet/wallet.h`

In commit _"build: enable external signer by default"_ (f24268f5fbf37de8a3716bd862f86dd7b831e80d) the `--enable-external-signer` configure option could be dropped in the `00_setup_env_native_multiprocess.sh` as well.

Not related to this PR, but is there any reasons why `NetworkArg` and `m_chain` are not private in the `ExternalSigner` class?

meshcollider reviewed (commented):
utACK 718fdce93c83f971d5fc70fe53c74d8a2165381a

Sjors commented:
Thanks. Will address the above comments if this needs rebase / touching.

> Not related to this PR, but is there any reasons why `NetworkArg` and `m_chain` are not private in the `ExternalSigner` class?

Possibly because in earlier iterations I would use them directly.

promag reviewed (commented):
Tested ACK 718fdce93c83f971d5fc70fe53c74d8a2165381a.

nit, since `walletdisplayaddress` is not included when external signer is disabled then it could also hide verify button with:

```diff
diff --git a/src/qt/receiverequestdialog.cpp b/src/qt/receiverequestdialog.cpp
index abe7de8f89..015b6c0b8d 100644
--- a/src/qt/receiverequestdialog.cpp
+++ b/src/qt/receiverequestdialog.cpp
@@ -11,6 +11,9 @@
 #include <qt/qrimagewidget.h>
 #include <qt/walletmodel.h>

+#include <interfaces/node.h>
+#include <interfaces/wallet.h>
+
 #include <QDialog>
 #include <QString>

@@ -90,7 +93,7 @@ void ReceiveRequestDialog::setInfo(const SendCoinsRecipient &_info)
         ui->wallet_content->hide();
     }

-    ui->btnVerify->setVisible(this->model->wallet().hasExternalSigner());
+    ui->btnVerify->setVisible(!model->node().externalSigners().empty() && this->model->wallet().hasExternalSigner());

     connect(ui->btnVerify, &QPushButton::clicked, [this] {
         model->displayAddress(info.address.toStdString());
```

achow101 commented:
ACK 718fdce93c83f971d5fc70fe53c74d8a2165381a

fanquake reviewed (changes requested):
Can you clean this up. The (re)-review burden here is basically non-existent, so I don't understand why obvious issues wouldn't be addressed. I'm also not sure what having to rebase (something I don't think should be an issue for contributors) has to do with fixing anything.

> Thanks. Will address the above comments if this needs rebase / touching.

This does need touching, because as pointed out by [that comment](https://github.com/bitcoin/bitcoin/pull/21935#pullrequestreview-682483034), the changes here are incomplete. If the bar for "retouching" isn't fixing a commit so that it's actually complete / correct, then what is it? 

Besides the current review comments, some other things to consider:

(depends build) Boost Process seems to use Boost ASIO, which uses `std::allocator<void>`, which was deprecated in C++17. This leads to a nice stack of warnings, at least on macOS:
```bash
bitcoin/depends/x86_64-apple-darwin20.5.0/include/boost/asio/associated_allocator.hpp:78:49: warning: 'allocator<void>' is deprecated [-Wdeprecated-declarations]
template <typename T, typename Allocator = std::allocator<void> >
                                                ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/memory:730:28: note: 'allocator<void>' has been explicitly marked deprecated here
class _LIBCPP_TEMPLATE_VIS _LIBCPP_DEPRECATED_IN_CXX17 allocator<void>
                           ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__config:1058:39: note: expanded from macro '_LIBCPP_DEPRECATED_IN_CXX17'
#  define _LIBCPP_DEPRECATED_IN_CXX17 _LIBCPP_DEPRECATED
                                      ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__config:1035:48: note: expanded from macro '_LIBCPP_DEPRECATED'
#    define _LIBCPP_DEPRECATED __attribute__ ((deprecated))
```

I'm also seeing unused-result warnings (from Boost) when building on my Ubuntu machine:
```bash
In file included from /home/ubuntu/bitcoin/depends/x86_64-pc-linux-gnu/include/boost/process/detail/execute_impl.hpp:24,
                 from /home/ubuntu/bitcoin/depends/x86_64-pc-linux-gnu/include/boost/process/child.hpp:22,
                 from /home/ubuntu/bitcoin/depends/x86_64-pc-linux-gnu/include/boost/process/async_system.hpp:22,
                 from /home/ubuntu/bitcoin/depends/x86_64-pc-linux-gnu/include/boost/process.hpp:24,
                 from util/system.cpp:9:
/home/ubuntu/bitcoin/depends/x86_64-pc-linux-gnu/include/boost/process/detail/posix/executor.hpp: In member function ‘void boost::process::detail::posix::executor<Sequence>::write_error(const std::error_code&, const char*) [with Sequence = boost::fusion::joint_view<boost::fusion::tuple<boost::process::detail::posix::exe_cmd_init<char> >, boost::fusion::filter_view<const boost::fusion::tuple<const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, boost::process::detail::posix::pipe_out<1, -1>&, boost::process::detail::posix::pipe_out<2, -1>&, boost::process::detail::posix::pipe_in&>, boost::process::detail::is_initializer<mpl_::arg<-1> > > >]’:
/home/ubuntu/bitcoin/depends/x86_64-pc-linux-gnu/include/boost/process/detail/posix/executor.hpp:156:16: warning: ignoring return value of ‘ssize_t write(int, const void*, size_t)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
  156 |         ::write(_pipe_sink, &len, sizeof(int));
      |         ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/ubuntu/bitcoin/depends/x86_64-pc-linux-gnu/include/boost/process/detail/posix/executor.hpp:159:16: warning: ignoring return value of ‘ssize_t write(int, const void*, size_t)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
  159 |         ::write(_pipe_sink, &len, sizeof(int));
      |         ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/ubuntu/bitcoin/depends/x86_64-pc-linux-gnu/include/boost/process/detail/posix/executor.hpp:160:16: warning: ignoring return value of ‘ssize_t write(int, const void*, size_t)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
  160 |         ::write(_pipe_sink, msg, len);
      |         ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
  AR       libbitcoin_util.a
```

Also, it looks like a bunch of review comments from #15382 were either just never implemented / followed up on. Any reason for that?

Sjors commented on src/Makefile.test.include:
```diff
@@ -161,6 +161,7 @@ BITCOIN_TESTS += \
 
 FUZZ_SUITE_LD_COMMON +=\
  $(LIBBITCOIN_WALLET) \
+ $(LIBBITCOIN_COMMON) \
```
All I can do here is split it into a separate commit and explain that I have no idea why it works. The [error messages](https://github.com/bitcoin/bitcoin/pull/21935#issuecomment-858214155) that lead to this were utterly useless. I have no idea what code is broken, so no idea what needs repairing. I also never worked on fuzzer code, which doesn't help.

ryanofsky replied:
> All I can do here is split it into a separate commit and explain that I have no idea why it works. The [error messages](https://github.com/bitcoin/bitcoin/pull/21935#issuecomment-858214155) that lead to this were utterly useless. I have no idea what code is broken, so no idea what needs repairing. I also never worked on fuzzer code, which doesn't help.

This seems fine but I wonder what is confusing about the error messages. On the surface, the error messages seem to just be saying libbitcoin_wallet.a can't find a few external_signer.cpp symbols. Since the symbols are in libbitcoin_common.a, it makes sense that linking libbitcoin_common.a after libbitcoin_wallet.a would fix the error.

I'm also a little confused by fanquake's comment https://github.com/bitcoin/bitcoin/pull/21935/files#r651410470 about adding a new circular dependency. I don't doubt that this PR adds one, but I'm not sure I see what it is. Maybe the comment is just referring to updating the Makefile to reflect circular dependencies in the code, though.

achow101 replied:
It looks like the linker error is simply due to the linking order. Everywhere else we have `LIBBITCOIN_WALLET`, it is always before `LIBBITCOIN_COMMON`. But if you go up to where `FUZZ_SUITE_LD_COMMON` is first set, you see that we will end up having `LIBBITCOIN_COMMON` set before `LIBBITCOIN_WALLET` which means that the linker will have problems linking things common things that the wallet uses. Because the order is correct for the other targets, we only see a linker error for `test/fuzz/fuzz`.

I would consider the following diff to be more correct than currently implemented:
```diff
diff --git a/src/Makefile.test.include b/src/Makefile.test.include
index 157d8ee2e1..fc2fd80166 100644
--- a/src/Makefile.test.include
+++ b/src/Makefile.test.include
@@ -35,11 +35,12 @@ BITCOIN_TEST_SUITE = \
   $(TEST_UTIL_H)
 
 FUZZ_SUITE_LD_COMMON = \
+ $(LIBTEST_UTIL) \
+ $(LIBTEST_FUZZ) \
  $(LIBBITCOIN_SERVER) \
+ $(LIBBITCOIN_WALLET) \
  $(LIBBITCOIN_COMMON) \
  $(LIBBITCOIN_UTIL) \
- $(LIBTEST_UTIL) \
- $(LIBTEST_FUZZ) \
  $(LIBBITCOIN_CONSENSUS) \
  $(LIBBITCOIN_CRYPTO) \
  $(LIBBITCOIN_CLI) \
@@ -159,12 +160,7 @@ BITCOIN_TESTS += \
   wallet/test/ismine_tests.cpp \
   wallet/test/scriptpubkeyman_tests.cpp
 
-# This adds libbitcoin again to fix linking external_signer_scriptpubkeyman.cpp
-# when configured with --disable-external-signer. It is unclear why this is
-# needed and may indicate a bug elsewhere.
 FUZZ_SUITE_LD_COMMON +=\
- $(LIBBITCOIN_WALLET) \
- $(LIBBITCOIN_COMMON) \
  $(SQLITE_LIBS) \
  $(BDB_LIBS)
 
```

In this diff, `LIBTEST_UTIL` and `LIBTEST_FUZZ` are moved to the top because they include `LIBBITCOIN_SERVER` and `LIBBITCOIN_COMMON`. `LIBBITCOIN_SERVER` always needs to be the first item in the linker order since it has the most dependencies.

I think, in general, the makefiles for making the fuzz and test binaries should be revisited so that the linking order is made consistent with the rest of the code and to avoid other linker order issues that may crop up in the future.

Sjors commented:
> Also, it looks like a bunch of review comments from #15382 were either just never implemented / followed up on. Any reason for that?

I attempted to address several in #20614, but that was a dead end. Other things were done in master. Is there anything specific you're missing?

I'll look into the above issues meanwhile. Some of that seems to have been missed in bitcoin-core/gui#4 , but now is a good time to fix things.

Sjors committed:
fuzz: fix fuzz binary linking order

We encountered a linking error when attempting to include external_signer_scriptpubkeyman.cpp when configured with --disable-external-signer.

Everywhere else we have LIBBITCOIN_WALLET, it is always before LIBBITCOIN_COMMON. But if you go up to where FUZZ_SUITE_LD_COMMON is first set, you see that we will end up having LIBBITCOIN_COMMON set before LIBBITCOIN_WALLET which means that the linker will have problems linking things common things that the wallet uses. Because the order is correct for the other targets, we only see a linker error for test/fuzz/fuzz.

In this diff, LIBTEST_UTIL and LIBTEST_FUZZ are moved to the top because they include LIBBITCOIN_SERVER and LIBBITCOIN_COMMON. LIBBITCOIN_SERVER always needs to be the first item in the linker order since it has the most dependencies.

The makefiles for making the fuzz and test binaries should be revisited so that the linking order is made consistent with the rest of the code and to avoid other linker order issues that may crop up in the future.

Co-Authored-By: Andrew Chow <achow101-github@achow101.com>

ryanofsky reviewed (commented):
re: https://github.com/bitcoin/bitcoin/pull/21935#issue-643455659

> This adds 3 commits on top of [bitcoin-core/gui#4](https://github.com/bitcoin-core/gui/pull/4) and reverts #22173.

Does it say anywhere what motivation is for reverting #22173? It looks like it is no longer an error to load a wallet using an external signer when there is no external signer support compiled in. It seems like a good change, but maybe release notes or some description of the behavior change if there is one would be good. Is this the part of the PR that fixes #21919?

> I don't think we should expect GUI users to self compile.
> 
> For manual `bitcoind` builds this feature is still opt-in.

Is this still true? It seems like external signer default is just changed from no to yes for the whole build. Is the option applied differently to different binaries somehow?

> Side-note: this PR may or may not (currently) break CI for the GUI repository, as explained here: [bitcoin-core/gui#4 (comment)](https://github.com/bitcoin-core/gui/pull/4#issuecomment-769859001)

Is this still true?

Sjors committed:
refactor: make ExternalSigner NetworkArg() and m_chain private

Sjors commented:
Rebased and addressed some of the above comments. Will continue shortly with the `std::allocator` issue and the display address button.

ryanofsky reviewed (approved):
Code review ACK 718fdce93c83f971d5fc70fe53c74d8a2165381a. Great to see this feature enabled by default, and to be able to flexibly load wallets without the signer. Nice to remove a bunch of ifdef's in the code, too!

Agree with fanquake moving up the dependency tweak to a different commit would be nice too, because we will probably revisit that at some point, and a separate commit would make the origin less mysterious.

ryanofsky reviewed (approved):
Code review ACK d0ce4f2c6098761b3076d64ea65e37f1950cf085. Changes since last review: rebase, splitting commit fixing fuzz link error, adding commit cleaning up externalsigner class, and some tweaked includes

Sjors commented:
@promag I'm not sure what change you're proposing in https://github.com/bitcoin/bitcoin/pull/21935#pullrequestreview-682882043, but note that `model->node().externalSigners()` can be slow, since it calls `hwi.py enumerate` and the result is currently not cached.

@ryanofsky I'll ad a clarification to the revert commit. Basically as of 9599db6 loading an external signer wallet will just work, until you actually try to use the external signer. The commit describes the behaviour change, but I'll check if I can split it.

Sjors commented:
>> For manual bitcoind builds this feature is still opt-in.

> Is this still true? It seems like external signer default is just changed from no to yes for the whole build. Is the option applied differently to different binaries somehow?

You're right, in an earlier incarnation of https://github.com/bitcoin-core/gui/pull/4 `configure.ac` would turn external signer support on by default if the GUI was compiled. That was lost, either in a rebase or because it was too complicated.

I updated the title reflect this change. Alternatively I can try to recover how I did it before.

Regarding the impact of this PR on the GUI repo, I don't know, it depends on if @jonasschnelli changed the Win64 CI machine to match (or actually use) `00_setup_env_win64.sh`. See https://github.com/bitcoin-core/gui/pull/4#issuecomment-769859001 (but it's a trivial fix in any case).

@fanquake regarding `std::allocator<void>` deprecation and its use in Boost::Process, that's quite frustrating. Do we know if newer versions of Boost fix this?

Sjors committed:
gui: misc external signer fixes and translation hints

promag commented:
> @promag I'm not sure what change you're proposing in [#21935 (review)](https://github.com/bitcoin/bitcoin/pull/21935#pullrequestreview-682882043), but note that `model->node().externalSigners()` can be slow, since it calls `hwi.py enumerate` and the result is currently not cached.

Right! Anyway, not a problem now that #22173 is not reverted.

Sjors commented:
I added one commit that addresses a bunch of small issues raised in the GUI PR, mostly adding translation hints, and disabling some fields when support is not compiled.

Rspigler commented:
Was [this](https://github.com/bitcoin/bitcoin/pull/15382#discussion_r464500701) ever fixed?

achow101 commented:
ACK 2f5bdcbc31a2eeb7c11226a9e51c56f02ac807dd

hebasto reviewed (approved):
re-ACK 2f5bdcbc31a2eeb7c11226a9e51c56f02ac807dd

fanquake commented:
Thanks @ryanofsky, @achow101, @hebasto, @promag, @Rspigler and everyone else for your input here.

Rspigler commented:
Is there still a CI issue?

jonatack commented:
> Is there still a CI issue?

Not sure it's related but the https://bitcoinbuilds.org Win64 job has been failing for a few days now with *configure: error: Boost::Process is required for external signer support, but not available!*

Sjors commented:
@Rspigler @jonatack very likely related; the GUI build settings are hardcoded rather than using the config in ci: https://github.com/bitcoin-core/gui/pull/4#issuecomment-769859001

Rspigler commented:
I don't know how I feel about this being merged when there is still an outstanding CI issue, it introduces a new circular dependency, and there are some serious upstream boost::process [issues](https://github.com/bitcoin/bitcoin/pull/15382#discussion_r465212138) and unresolved [followups](https://github.com/bitcoin/bitcoin/pull/20614)

Sjors commented:
@Rspigler the GUI CI is misconfigured, which is unrelated to this PR (and trivial to fix by whoever has access to it).

The PR no longer introduces a circular dependency (see fc0eca31b33f87882e2aa329a3746d4e08af1985).

It makes sense to emphasise that this is still experimental (descriptor wallets already are, which this builds on). But the Boost::Process code is unused if you don't configure `-signer=`, so I don't think that should block enabling the code. If you point to a malicious Python script, bad things can happen anyway, because that script runs as the same user.
